src/decoder.rs

```diff
@@ -46,8 +46,6 @@ pub struct Decoder<R> {
 }
 
 impl<R> Decoder<R>
-where
-    R: Read,
```

Comment:
Why can we now publicly instantiate the `Decoder` with something that is not `Read`? Doesn't that push the possible compile error to a point further in the code that what actually causes it and make it harder to understand and thus fix?

Reply:
It's a quite common pattern I've seen others in rust also. In the struct declaration, you don't limit the type parameters, but rather put limits on the impl.

```rust
struct MyFineType<T> {
  ...
}

impl<T: Read> MyFineType<T> {
  pub new(t: T) -> Self { … }
}
```

For a struct with private fields, it wouldn't make any difference since you can't reach `new()` without implementing Read.

I guess it's just less to write, especially if you have a complicated trait bound.

Reply:
I've indeed seen it be useful to lift these trait bounds from the `struct` when describing complex trait bounds on your struct: if the requirements on for `Decoder<R>: Read` were more than `R: Read ` and were, say `R: Read + T1<A,B> + T2<C,D> where C: T3`, then this writing enable writing only a `where Decoder<R>: Read` instead of all those bounds.

But in our case, `fn f<R>() where Decoder<R>: Read` is actually much longer to write than `fn f<R: Read>()`, so I don't see that gain here. On a similar question, the standard library has chosen to not lift it: https://doc.rust-lang.org/std/io/struct.BufReader.html#impl

But anyway, my point was about a slightly different problem: it was already lifted from the `struct`, and currently this pull request is lifting it from the `new` impl, unlike the example you've provided, so we *can* reach `new()` without implementing Read, hence the ability to "publicly instantiate the `Decoder` with something that is not `Read`" despite the fact that the fields are private, and the "push[ing] the possible compile error to a point further in the code than what actually causes it".

Reply:
You're way ahead of me. Good point! :D

Reply:
I had no particular grand plan in putting them methods outside of the bound, so I've reverted them back within the bounds based on this feedback in the latest commit. Thanks all!